streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -84,6 +84,8 @@
     private Producer<byte[], byte[]> producer;
     private boolean commitRequested = false;
     private boolean transactionInFlight = false;
+    private Map<TopicPartition, Long> partitionTimes = new HashMap<>();
+    private boolean isCloseTriggerCommit = false;
```

Comment:
I am not sure why we need this variable? Can you elaborate?

Reply:
Oh, because we need to differentiate between a commit that is triggered by a regular process or by a close. If we call partitionTime() in a commit triggered by a close() call, then partitionTime() would always return -1. (recall that due to the order of operations in close, the partition times has been reset to -1 first before the commit call was made).

Reply:
Thanks. Understood.

It might be better, to actually change `Stream#commit(boolean startNewTransaction)` to accept a second parameter `Map<TopicPartition, Long> partitionTimes` to pass in the information.

In `close()` before we actually "loose" the timestamps we preserve them and pass into `commit()` later. In a regular `commit()` we get the timestamps from the `partitionGroup` (ie, some code that is now in `commit(boolean)` would go into `commit()`).

This would avoid the requirement to introduce the flag and make the code more readable, because decision are more local an encapsulated in each method without cross-method dependencies.

Reply:
Oh, sure, that would work.